File: 566.Reshape-the-Matrix.py
# https://leetcode.com/problems/reshape-the-matrix/
#
# algorithms
# Easy (58.16%)
# Total Accepted:    67,646
# Total Submissions: 116,319

class Solution(object):
    def matrixReshape(self, nums, r, c):
        """
        :type nums: List[List[int]]
        :type r: int
        :type c: int
        :rtype: List[List[int]]
        """
        row, col = len(nums), len(nums[0])
        
        if r * c != row * col:
            return nums
        
        res = [[0] * c for _ in xrange(r)]
        old_i, old_j = 0, 0
        
        for i in xrange(r):
            for j in xrange(c):
                if old_j == col:
                    old_i += 1
                    old_j = 0
                res[i][j] = nums[old_i][old_j]
                old_j += 1
                
        return res
    

File: 566.Reshape-the-Matrix.java
// https://leetcode.com/problems/reshape-the-matrix/
//
// algorithms
// Easy (47.30%)
// Total Accepted: 92,216
// Total Submissions: 153,991
// beats 100.0% of java submissions


class Solution {
    public int[][] matrixReshape(int[][] nums, int r, int c) {
        int m = nums.length;
        int n = nums[0].length;

        if (r * c != m * n) {
            return nums;
        }

        int[][] res = new int[r][c];
        int xIdx = 0, yIdx = 0;
        for (int i = 0; i < r; i++) {
            for (int j = 0; j < c; j++) {
                res[i][j] = nums[xIdx][yIdx];

                yIdx++;
                if (yIdx == n) {
                    xIdx++;
                    yIdx = 0;
                }
            }
        }

        return res;
    }
}